src/chainstate/stacks/auth.rs

```diff
@@ -21,44 +21,1431 @@ use net::StacksMessageCodec;
+                let mut pubkey = pubkey_buf.to_public_key()?;
+                pubkey.set_compressed(false);
+                
+                TransactionAuthField::PublicKey(pubkey)
+            },
+            x if x == TransactionAuthFieldID::SignatureCompressed as u8 => {
+                let sig : MessageSignature = read_next(buf, &mut index, max_size)?;
+                TransactionAuthField::Signature(TransactionPublicKeyEncoding::Compressed, sig)
+            },
+            x if x == TransactionAuthFieldID::SignatureUncompressed as u8 => {
+                let sig : MessageSignature = read_next(buf, &mut index, max_size)?;
+                TransactionAuthField::Signature(TransactionPublicKeyEncoding::Uncompressed, sig)
+            },
+            _ => {
+                test_debug!("Failed to deserialize auth field ID {}", field_id);
+                return Err(net_error::DeserializeError);
+            }
+        };
 
+        *index_ptr = index;
+        Ok(field)
+    }
+}
+
+impl StacksMessageCodec for MultisigSpendingCondition {
+    fn serialize(&self) -> Vec<u8> {
+        let mut res = vec![];
+       
+        write_next(&mut res, &(self.hash_mode.clone() as u8));
+        write_next(&mut res, &self.signer);
         write_next(&mut res, &self.nonce);
-        write_next(&mut res, &public_key_buffers);
-        write_next(&mut res, &self.signatures);
+        write_next(&mut res, &self.fields);
         write_next(&mut res, &self.signatures_required);
+        
         res
     }
 
-    fn deserialize(buf: &Vec<u8>, index: &mut u32, max_size: u32) -> Result<TransactionAuth, net_error> {
-        let nonce : u64                                     = read_next(buf, index, max_size)?;
-        let public_key_buffers: Vec<StacksPublicKeyBuffer>  = read_next(buf, index, max_size)?;
-        let signatures: Vec<MessageSignature>               = read_next(buf, index, max_size)?;
-        let signatures_required: u16                        = read_next(buf, index, max_size)?;
+    fn deserialize(buf: &Vec<u8>, index_ptr: &mut u32, max_size: u32) -> Result<MultisigSpendingCondition, net_error> {
+        let mut index = *index_ptr;
+
+        let hash_mode_u8 : u8 = read_next(buf, &mut index, max_size)?;
+        let hash_mode = MultisigHashMode::from_u8(hash_mode_u8)
+            .ok_or(net_error::DeserializeError)?;
+
+        let signer : Hash160 = read_next(buf, &mut index, max_size)?;
+        let nonce : u64 = read_next(buf, &mut index, max_size)?;
 
-        // attempt to parse all public keys
-        let mut public_keys = vec![];
-        for pubkey_buf in &public_key_buffers {
-            let pubk = pubkey_buf.to_public_key()?;
-            public_keys.push(pubk);
+        let fields : Vec<TransactionAuthField> = read_next(buf, &mut index, max_size)?;
+        let signatures_required: u16 = read_next(buf, &mut index, max_size)?;
+        
+        // read and decode _exactly_ num_signatures signature buffers
```

Comment:
this comment doesn't appear accurate -- it looks like line 151 will parse however many TransactionAuthFields are supplied in the input

Reply:
A `TransactionAuthField` can be a public key or a signature.  The number of signatures required can be less than the number of auth fields.  Right now, there's no limit on how many auth fields a transaction can have; only that `num_signatures` must match the number of `TransactionAuthField::Signature`s present in the `fields` vector.  I expect that there will be a soft-fork default limit to the number of fields, but I'm not sure it's a good idea to force an upper bound at the protocol level (miners will figure that out by themselves).

Reply:
Gotcha -- that's fine by me. I think the comment led me to believe the function should be failing fast if more signatures are read than `signatures_required`, but that isn't really necessary.